File: 03_Stack/04_Generate_Parentheses.py
class Solution(object):
    
    def genSequence(self, ans, currString, openCount, closed, n, currPar):
        currString += currPar
        if currPar == "(":
            openCount += 1
        else:
            closed += 1
        
        if openCount == n and closed == n:
            ans.append(currString)
        
        if openCount < n:
            self.genSequence(ans, currString, openCount, closed, n, "(")
        if closed < openCount:
            self.genSequence(ans, currString, openCount, closed, n, ")")
    
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        ans = []
        currString = ""

        self.genSequence(ans, currString, 0, 0, n, "(")
        return ans

File: 03_Stack/04_Generate_Parentheses.cpp
class Solution
{
public:
    void genSequence(vector<string> &ans, string currString, int open, int closed, int n, string currPar)
    {
        currString += currPar;
        if (currPar == "(")
            open++;
        else
            closed++;

        if (open == n && closed == n)
        {
            ans.push_back(currString);
            return;
        }

        if (open < n)
            genSequence(ans, currString, open, closed, n, "(");
        if (closed < open)
            genSequence(ans, currString, open, closed, n, ")");
    }

    vector<string> generateParenthesis(int n)
    {
        vector<string> ans;
        string currString = "";
        genSequence(ans, currString, 0, 0, n, "(");
        return ans;
    }
};